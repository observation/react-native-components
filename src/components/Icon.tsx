import React from 'react'

import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome'

import Icons, { IconName } from '../lib/Icons'
import theme from '../styles/theme'

export type IconStyleProp = 'light' | 'solid'

type IconAppearanceProps = {
  style?: IconStyleProp
  color?: string
  size?: number
  testID?: string
}

type IconProps = IconAppearanceProps & {
  name: IconName
}

export const Icon = ({ name, color, size, testID, style }: IconProps): JSX.Element => {
  const iconStyle = style ?? 'light'
  const icon = iconStyle === 'light' ? Icons[name].light : Icons[name].solid
  const iconColor = color ?? theme.color.primary
  const iconSize = size ?? theme.icon.size.large

  return <FontAwesomeIcon icon={icon} color={iconColor} size={iconSize} testID={testID} />
}
