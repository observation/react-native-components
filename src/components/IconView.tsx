import React from 'react'
import { StyleProp, StyleSheet, View, ViewStyle } from 'react-native'

import theme from '../styles/theme'

type Props = {
  icon?: JSX.Element
  lineHeight?: number
  children?: React.ReactNode
  containerStyle?: StyleProp<ViewStyle>
  iconContainerStyle?: StyleProp<ViewStyle>
}

const IconView = ({ icon, lineHeight = theme.margin.large, children, containerStyle, iconContainerStyle }: Props) => (
  <View style={[styles.containerStyle, containerStyle]}>
    {icon && (
      <View style={[styles.iconContainer, { width: lineHeight }, iconContainerStyle]}>
        <View style={[styles.iconInnerContainer, { height: lineHeight }]}>{icon}</View>
      </View>
    )}
    <View style={{ flexShrink: 1 }}>{children}</View>
  </View>
)

export default IconView

const styles = StyleSheet.create({
  containerStyle: {
    flexDirection: 'row',
  },
  iconContainer: {
    alignItems: 'flex-start',
  },
  iconInnerContainer: {
    justifyContent: 'center',
  },
})
