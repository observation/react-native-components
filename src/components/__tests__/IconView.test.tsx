import React from 'react'
import { Text } from 'react-native'

import { render } from '@testing-library/react-native'

import { IconInfo } from '../Icon'
import IconView from '../IconView'

describe('IconView', () => {
  describe('Rendering', () => {
    test('With default parameters', () => {
      const { toJSON } = render(
        <IconView>
          <Text>Text to show</Text>
        </IconView>,
      )
      expect(toJSON()).toMatchSnapshot()
    })

    test('With specified parameters', () => {
      const { toJSON } = render(
        <IconView containerStyle={{ flex: 1 }} iconContainerStyle={{ flex: 1 }} icon={<IconInfo />} lineHeight={14}>
          <Text>Text to show</Text>
        </IconView>,
      )
      expect(toJSON()).toMatchSnapshot()
    })
  })
})
