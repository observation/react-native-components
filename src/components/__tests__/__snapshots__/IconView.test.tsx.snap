// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconView Rendering With default parameters 1`] = `
<View
  style={
    [
      {
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "flexShrink": 1,
      }
    }
  >
    <Text>
      Text to show
    </Text>
  </View>
</View>
`;

exports[`IconView Rendering With specified parameters 1`] = `
<View
  style={
    [
      {
        "flexDirection": "row",
      },
      {
        "flex": 1,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "alignItems": "flex-start",
        },
        {
          "width": 14,
        },
        {
          "flex": 1,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "justifyContent": "center",
          },
          {
            "height": 14,
          },
        ]
      }
    >
      <Icon
        color="#0066B1"
        name="circle-info"
        size={16}
        type="light"
      />
    </View>
  </View>
  <View
    style={
      {
        "flexShrink": 1,
      }
    }
  >
    <Text>
      Text to show
    </Text>
  </View>
</View>
`;
